Feature/ai builder connector

AdrianAlcolea opened the pull request:
## Change
New connector for the AIBuilder platform

## How to Test
There are unitary test created for it.
To test the actual functionality you need to add a valid API_TOKEN (from a user of the platform) to the 'override.env' file and use the 'aibuilder' profile when launching docker compose. Then you will see new ml_models added to the platform (you can search selecting the 'aibuilder' platform).

## Checklist
The only thing I didn't do is to add an explanation on the README (in general I see that there are no such explanations for the use of other connectors, but maybe the TOKEN thing should be commented). I guess we can discuss this on the PR and decide if it needs some explanation.

AdrianAlcolea committed:
Merged with develop

AdrianAlcolea committed:
Structure with empty token

AdrianAlcolea committed:
Merging with develop

AdrianAlcolea committed:
Wotking with hidden token

AdrianAlcolea committed:
Infinite loop on empty lists solved

AdrianAlcolea committed:
Merged with develop

AdrianAlcolea committed:
Happy path test of aibuilder connector

AdrianAlcolea committed:
First working test for aibuilder connector

AdrianAlcolea committed:
Merged with develop

AdrianAlcolea committed:
Catalog list tests done

AdrianAlcolea committed:
All AIBuilder Connector test passed

AdrianAlcolea committed:
Merged with develop

AdrianAlcolea committed:
TODO removed. Continuation after catalog error is not needed.

AdrianAlcolea committed:
Changed the venv name and removed the venv_aiod directory

AdrianAlcolea committed:
Added aibuilder to the example for usin connectors on the README

AdrianAlcolea committed:
Documentation texts added to the aibuilder connector class and methods

AdrianAlcolea committed:
Modified same_as field for security reasons

AdrianAlcolea committed:
Comment about same_as field aded to the README file

AdrianAlcolea committed:
Comment about same_as field aded to the README file

AdrianAlcolea committed:
Comment about same_as field aded to the README file

PGijsbers commented on .env:
```diff
@@ -20,6 +20,9 @@ AIOD_KEYCLOAK_PORT=8080
 EGICHECKINALIAS=
 
+#AIBUILDER
+API_TOKEN=""
```
Probably better to update this to `AIBUILDER_API_TOKEN`, there are likely going to be more API tokens.

PGijsbers commented on connectors/aibuilder/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY mlmodels.sh /opt/connectors/script/mlmodels.sh
+COPY entry.sh /opt/connectors/script/entry.sh
+
+USER root
```
```suggestion
USER root
RUN apt-get update
```

cron was not installable for me otherwise

PGijsbers commented on docker-compose.yaml:
```diff
@@ -109,6 +109,27 @@ services:
+    volumes:
+      - ./src:/app
+      - ${DATA_PATH}/connectors:/opt/connectors/data
+      - ./connectors/aibuilder/:/opt/connectors/script
```
Please include the default scripts directly in the image, optionally provide mounted scripts for live updates in the `docker-compose.dev.yaml`. This makes it so that when running without `USE_LOCAL_DEV`, you do not accidentally include local changes (provided images were built with a clean slate). Mounting the data directory (for writing logs) can stay, of course. See for example the zenodo connector.

PGijsbers commented on connectors/aibuilder/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM aiod_metadata_catalogue
+
+COPY cron /etc/cron.d/aiod
+COPY mlmodels.sh /opt/connectors/script/mlmodels.sh
+COPY entry.sh /opt/connectors/script/entry.sh
```
Move these commands to below the installation of cron. That way, when rebuilding the image when these files are updated, you can use your build cache to avoid having to re-download/install cron.

PGijsbers commented on src/connectors/aibuilder/aibuilder_mlmodel_connector.py:
```diff
@@ -0,0 +1,292 @@
+
+from .aibuilder_mappings import mlmodel_mapping
+
+TOKEN = os.getenv("API_TOKEN", "")
```
Please add an immediate and clear error when the `API_TOKEN` is not set.

PGijsbers commented on src/connectors/aibuilder/aibuilder_mlmodel_connector.py:
```diff
@@ -0,0 +1,292 @@
+        response = requests.get(url, timeout=REQUEST_TIMEOUT)
+        if not response.ok:
+            status_code = response.status_code
+            msg = response.json()["error"]["message"]
```
This errors when the content of the response is not JSON. I found that for the 401: UNAUTHORIZED response, the server does not return JSON, so the connector would crash.

AdrianAlcolea replied:
I added a try - except block returning a record error (it is maybe too generic error for that, but I think it is the proper way in this context to return it)

PGijsbers replied:
The problem is not with the request itself (though I guess there can also be errors there). The request is successful but it will have status `401: UNAUTHORIZED` and content `b'authentication: null'` or something similar. The latter is not JSON, so the `response.json()` call below will result in an error.

AdrianAlcolea replied:
Ah, ok. I took this behaviour from the openml connectors, so maybe we should revise them too. I am looking now at the zenodo one and I will change the approach to that one

PGijsbers reviewed (changes requested):
Don't have it working yet (probably misconfigured my API_TOKEN?), so have to try that tomorrow. Here's already some changes though. Assuming it otherwise works I think that's all that really needs to be addressed.

AdrianAlcolea committed:
Modifications for the Pull Request

AdrianAlcolea commented:
I made all the modifications, also added an __init__ method to be able of passing a non-empty TOKEN for the tests.

AdrianAlcolea committed:
Removed extra set_token method

PGijsbers commented on docker-compose.yaml:
```diff
@@ -109,6 +109,26 @@ services:
+      dockerfile: Dockerfile
+    image: aiod_aibuilder_connector
+    container_name: aibuilder-connector
+    env_file: .env
```
Hm. I thought I left a comment about this earlier. It should not be necessary to forward the entire env file to the docker container. you should be able to remove that line. If things break, then you'd need to see which environment variable is missing and forward it with "environment" instead.

AdrianAlcolea replied:
Oh, that's true. I started it before the 'override.env' modification =P

AdrianAlcolea committed:
Handle successfull requests without json response. Also specific message for 401 unauthorized.

AdrianAlcolea committed:
Added unitary test for Unauthorized token error.

AdrianAlcolea committed:
Removed .env file from docker_compose

AdrianAlcolea commented:
I modified the code to handle the responses that are correct but not a 200, added an specific error message for the 401 error and made a new unitary test for it. Also removed the .env file from docker compose.

AdrianAlcolea committed:
Merged with develop

PGijsbers reviewed (approved):
LGTM
